clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java

```diff
@@ -1126,9 +1136,10 @@ public void onFailure(RuntimeException e) {
                                             // as the duration of the rebalance timeout. If we stop sending heartbeats,
                                             // however, then the session timeout may expire before we can rejoin.
                                             heartbeat.receiveHeartbeat();
+                                        } else if (e instanceof FencedInstanceIdException) {
+                                            heartbeatThread.failed.set(e);
```

Comment:
Maybe worth a log message that we caught the fenced error in the heartbeat thread.